### PR TITLE
Manage newsletter subscription in a modern ajax way

### PR DESCRIPTION
**Pull request by jf-viguier**
This is more elegant and avoid full page reload, see https://github.com/PrestaShop/PrestaShop/issues/17054

Newsletter subscription uses an old fashion "post to home page" form with several issues :
- After subscribing, the user is redirected to the home page.
- If the module is hooked several time, a buggy message tel the user that his email has already been used (see https://github.com/PrestaShop/PrestaShop/issues/15629)

Notice : I've changed the version number to 2.5.2 to launch upgrade : I must register a new hook actionFrontControllerSetMedia to load my new js file.

### CODE REVIEW

**Comment by jf-viguier** [comment]
I've made a new commit to replace ajaxRender by ajaxDie for Prestashop < 1.7.5 support

File: views/js/ps_emailsubscription.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+*  @copyright 2007-2020 PrestaShop SA
+*  @license   http://opensource.org/licenses/afl-3.0.php  Academic Free License (AFL 3.0)
+*  International Registered Trademark & Property of PrestaShop SA
+*
+* Don't forget to prefix your containers with your own identifier
+* to avoid any conflicts with others containers.
+*/
+
+$(document).ready(function () {
+    $('.block_newsletter form').on("submit", function () {
```
Comment (PierreRambaud) [inline views/js/ps_emailsubscription.js]:
```suggestion
    $('.block_newsletter form').on('submit', function () {
```

File: views/js/ps_emailsubscription.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        if (psemailsubscription_subscription !== 'undefined') {
```
Comment (PierreRambaud) [inline views/js/ps_emailsubscription.js]:
```suggestion
        if (psemailsubscription_subscription == undefined) {
          return true;
        }
```
Not sure it must be compared to a string

File: views/js/ps_emailsubscription.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            $('.block_newsletter_alert').remove();
+            $.ajax({
+                type: 'POST',
+                dataType: 'JSON',
+                url: psemailsubscription_subscription,
+                cache: false,
+                data: $(this).serialize(),
+                success: function (data) {
+                    if (data.nw_error) {
+                        psemailsubscriptionForm.prepend('<p class="alert alert-danger block_newsletter_alert">' + data.msg + '</p>');
```
Comment (PierreRambaud) [inline views/js/ps_emailsubscription.js]: use `showErrorMessage` and `showSuccessMessage` instead? :thinking:

File: views/templates/hook/ps_emailsubscription.tpl [hunk 4]
```diff
@@ -23,13 +23,13 @@
...
 
-<div class="email_subscription" id="blockEmailSubscription_{$hookName}">
+<div class="email_subscription block_newsletter" id="blockEmailSubscription_{$hookName}">
   <h4>{l s='Newsletter' d='Modules.Emailsubscription.Shop'}</h4>
   {if $msg}
     <p class="notification {if $nw_error}notification-error{else}notification-success{/if}">{$msg}</p>
   {/if}
   <form action="{$urls.current_url}#blockEmailSubscription_{$hookName}" method="post">
-    <input type="email" name="email" value="{$value}" placeholder="{l s='Your e-mail' d='Modules.Emailsubscription.Shop'}" required />
+    <input type="text" name="email" value="{$value}" placeholder="{l s='Your e-mail' d='Modules.Emailsubscription.Shop'}" />
```
Comment (PierreRambaud) [inline views/templates/hook/ps_emailsubscription.tpl]: Why? :thinking:

File: views/js/ps_emailsubscription.js (hunk 3 above)
Comment (Progi1984) [inline views/js/ps_emailsubscription.js]:
```suggestion
                        psemailsubscriptionForm.prepend(`<p class="alert alert-danger block_newsletter_alert">${data.msg}</p>`);
```

File: views/js/ps_emailsubscription.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+                    } else {
+                        psemailsubscriptionForm.prepend('<p class="alert alert-success block_newsletter_alert">' + data.msg + '</p>');
```
Comment (Progi1984) [inline views/js/ps_emailsubscription.js]:
```suggestion
                        psemailsubscriptionForm.prepend(`<p class="alert alert-success block_newsletter_alert">${data.msg}</p>`);
```

File: views/templates/hook/ps_emailsubscription.tpl (hunk 4 above)
Comment (jf-viguier) [reply views/templates/hook/ps_emailsubscription.tpl]: I've not changed anything here, anyway, the tpl in charge is in classic theme and is totally different.

File: views/js/ps_emailsubscription.js (hunk 2 above)
Comment (jf-viguier) [reply views/js/ps_emailsubscription.js]: My mistake, it's typeof psemailsubscription_subscription !== 'undefined'

File: views/js/ps_emailsubscription.js (hunk 3 above)
Comment (jf-viguier) [reply views/js/ps_emailsubscription.js]: Is only for bo, these are not available in front. I've stitched to non javascript behavior and alert messages in customer subscription/login.

Comment (jf-viguier) [reply views/js/ps_emailsubscription.js]: No,  template string is not supported by IE.
The entire module works in IE, I don't want to break it just for this.

File: views/js/ps_emailsubscription.js (hunk 5 above)
Comment (jf-viguier) [reply views/js/ps_emailsubscription.js]: No,  template string is not supported by IE.
The entire module works in IE, I don't want to break it just for this.

File: views/js/ps_emailsubscription.js (hunk 3 above)
Comment (PierreRambaud) [reply views/js/ps_emailsubscription.js]: Yes, we can use this kind of feature only when using webpack + babel :)

File: views/js/ps_emailsubscription.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+*  @license   http://opensource.org/licenses/afl-3.0.php  Academic Free License (AFL 3.0)
+*  International Registered Trademark & Property of PrestaShop SA
+*
+* Don't forget to prefix your containers with your own identifier
+* to avoid any conflicts with others containers.
+*/
+
+$(document).ready(function () {
+    $('.block_newsletter form').on('submit', function () {
+        if (typeof psemailsubscription_subscription !== 'undefined') {
```
Comment (PierreRambaud) [inline views/js/ps_emailsubscription.js]: Don't encapsulate the whole block, return as fast as possible.

File: views/js/ps_emailsubscription.js (hunk 3 above)
Comment (Progi1984) [reply views/js/ps_emailsubscription.js]: OK. Thanks ;)

File: views/js/ps_emailsubscription.js (hunk 6 above)
Comment (jf-viguier) [reply views/js/ps_emailsubscription.js]: ok

Comment (jf-viguier) [reply views/js/ps_emailsubscription.js]: done

**Review by PierreRambaud — DISMISSED** [review]

**Comment by jf-viguier** [comment]
Any news from QA ?

**Comment by jf-viguier** [comment]
Any news from QA ?

**Review by Progi1984 — DISMISSED** [review]

File: views/js/index.php [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+				    	
+header("Expires: Mon, 26 Jul 1997 05:00:00 GMT");
+header("Last-Modified: ".gmdate("D, d M Y H:i:s")." GMT");
+						
+header("Cache-Control: no-store, no-cache, must-revalidate");
+header("Cache-Control: post-check=0, pre-check=0", false);
+header("Pragma: no-cache");
+						
+header("Location: ../");
+exit;
```
Comment (NeOMakinG) [inline views/js/index.php]: Hey @jf-viguier 

Configure your IDE so it auto-add new line at the end of the file :)

Comment (NeOMakinG) [reply views/js/index.php]: Fix that asap and ping us we'll instant approve it

Thanks for your contribution anyway :)

**Review by NeOMakinG — DISMISSED** [review]

**Review by matks — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Comment by NeOMakinG** [comment]
Hey @jf-viguier

Please configure your IDE so it auto-add new line at the end of the file :)

**Comment by SD1982** [comment]
Hello @jf-viguier The success message when subscribing is not displayed, only the green rectangle appears .

Manage newsletter subscription in a modern ajax way #48 after PR

Can you check this please ? Other than that everything is OK

**Comment by jf-viguier** [comment]
@SD1982 Maybe a lang issue, what language do you use in front ? I didn't add any new message.
Do you have messages when javascript is turned off ?

**Comment by SD1982** [comment]
Hello @jf-viguier  I tried with the FO in English and in French same result. But by trying as you suggested by disabling javascript the success or failure messages are correctly displayed but this time they are below : 

**English with javascript disabled**

#48 after PR4

# 48 after PR5

**French  with javascript disabled**

# 48 after PR6

# 48 after PR7

**Comment by jf-viguier** [comment]
Hello @SD1982, I don't have this behavior. I'm in 1.7.6.2 and you ?
I've changed message position to match other ps behaviors like in account creation page.

**Comment by matks** [comment]
@jf-viguier I tested your PR also (on 1.7.7.x). I have the same issue:
Capture d’écran 2020-02-20 à 11 12 00

I checked the AJAX query, the `data.msg` data returned by backend is empty
Capture d’écran 2020-02-20 à 11 13 09

```
{"value":"mathieu.fermentb@prestashop.comsss","msg":"","conditions":"Vous pouvez vous d\u00e9sinscrire \u00e0 tout moment. Vous trouverez pour cela nos informations de contact dans les conditions d'utilisation du site."}
```

**Comment by SD1982** [comment]
Hello @jf-viguier  This behavior appears when i test on version 1.7.7 but I just tested on version 1.7.6 like you and indeed it works correctly and as you can see the message is diplayed on top this time  :

**Version 1.7.6.x**

#48 1 7 6 x after PR1

#48 1 7 6 x after PR2


 so it probably does not come from your PR what do you think of that @matks , @NeOMakinG

**Comment by jf-viguier** [comment]
@SD1982 please use a stable version ;-) You've certainly found a bug for 1.7.7 but I've no time to check this for now.
Certainly a bug in ajax request that doesn't have correct context of global vars stuff. @matks ? Other ajax messages works on 1.7.7 ?

**Comment by Robin-Fischer-PS** [comment]
I think we need @PrestaShop/prestashop-core-developers opinion on this ! If this PR works in 176 but not in 177, is this link to the PR or to the 177 version ?

Thanks

**Comment by jf-viguier** [comment]
@Robin-Fischer-PS I've already answered this : it's linked to 177 version. 
I just use old messages in ajax return array. 
Tt's certainly a bug in ajax request that doesn't have correct context of global vars stuff.

**Comment by matks** [comment]
> @Robin-Fischer-PS I've already answered this : it's linked to 177 version.
> I just use old messages in ajax return array.
> Tt's certainly a bug in ajax request that doesn't have correct context of global vars stuff.

It does not matter. We are not going to accept a Pull Request that _we know it will fail with 177_ 😉. What kind of professionnals would we be if we accept something that we know will fail with next release when it's scheduled for 1 or 2 months in the future ?

Until we know why it fails with 177 we cannot merge this PR.

It does not mean we will not merge this PR. It means we need more exploration.

**Comment by jf-viguier** [comment]
Thanks @matks  I understand. I just means that my pr is not the source of the bug. I guess that other ajax requests have this bug in 177 ?

**Comment by matks** [comment]
Found the culprit
https://github.com/PrestaShop/ps_emailsubscription/pull/37/files#diff-cfb1aa2304713070e111aaa57e5c8ef5R350

**Comment by matks** [comment]
@jf-viguier Fixed with commit https://github.com/PrestaShop/ps_emailsubscription/pull/48/commits/da17e14412f365758d7b63b6629595a42151330c

This PR can now go through QA again, for both 176 and 177 checks 💪

**Review by matks — APPROVED** [review]

**Review by NeOMakinG — APPROVED** [review]

**Comment by jf-viguier** [comment]
Thanks @matks for the fix, you rocks

**Review by PierreRambaud — APPROVED** [review]

**Comment by SD1982** [comment]
LGTM in 1.7.6.x and 1.7.7.x Thanks @jf-viguier

**Comment by Progi1984** [comment]
Thanks @jf-viguier

**Comment by jf-viguier** [comment]
@Progi1984 when this will pr be released ?

**Comment by jf-viguier** [comment]
Newsletter subscription seems no more in ajax in hummingbird theme. Can someone confirm ?

**Comment by wstoch** [comment]
I can confirm. Fresh install of hummingbird lacks 'block_newsletter' class in newsletter form template, so submit event isn't firing.

**Pull request merged**